### ISSUE TITLE
Request: support for VK_KHR_draw_indirect_count

### ISSUE THREAD

**Issue by Hindrik1997** [KhronosGroup/MoltenVK#168]
As of Vulkan 1.1.76 the VK_KHR_draw_indirect_count extension was added to the spec. The extension allows for the gpu to set the instance counts. Could this be added to MoltenVK?

**Comment by mellinoe** [comment]
Metal doesn't really support this (afaik), so it would have to be emulated in a way that most likely does not match up with the intent of the extension. By that, I mean it would need to read the draw count back to the CPU and then emit individual indirect draw calls -- nothing you couldn't already do without the extension.

**Comment by Hindrik1997** [comment]
In that case we will have to wait for Apple to add support for it in Metal.

**Comment by oscarbg** [comment]
On Twitter I said:

> Metal 2.1 adds indirect command buffers guess is similar to VK_NVX_device_generated_commands in Vulkan.. slightly related don't know if GPU setting draw counts in GPU buffers similar VK_KHR_draw_indirect_count support was already in Metal 2.0 or needs new 2.1 ICB's..


and got this reponse form @zeuxcg on Twitter: 

> This was there before 2.0 - unsure what version, https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515467-drawprimitives …. That’s only a single draw call (instanced) - MDI is only since 2.1 AFAIK.

**Comment by BeastLe9enD** [comment]
is there any news regarding the support of VK_KHR_draw_indirect_count ?

**Comment by MarloCraft** [comment]
Is here anybody who tried implementing this? Or is here anybody who plans to implement this in future ? :)

**Comment by alecazam** [comment]
Metal's draw indirect api is missing a drawCount, and only provides an indirectBufferOffet.   So you can't accumulate multi-material or multiple draws into a buffer, and then use the drawIndexPrimitive calls that take an indirectBuffer.  An offset is not a range unlike the DX12 and Vulkan implementation which are actually functional.

**Comment by fabiopolimeni** [comment]
> Metal's draw indirect api is missing a drawCount, and only provides an indirectBufferOffet. So you can't accumulate multi-material or multiple draws into a buffer, and then use the drawIndexPrimitive calls that take an indirectBuffer. An offset is not a range unlike the DX12 and Vulkan implementation which are actually functional.

@alecazam Isn't [Indirect Arguments](https://developer.apple.com/documentation/metal/mtldrawindexedprimitivesindirectarguments) + [Indirect Range Buffer](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/3020699-executecommandsinbuffer) sufficient for implementing VK_KHR_draw_indirect_count. It looks they are just made the same to me.

**Comment by alecazam** [comment]
It is nearly the same API for MDI on Metal with several big limiters.  indexStart should be int32_t, but Apple got that wrong.  There is also no stride support for storing a gl_drawID in the same MDI buffer.   

The problem happens when you try to call it.  You must have the cpu submit one draw call for each item in the buffer.  There is no indirect count, but the gpu can overwrite the instance count in this buffer.   So you would have to submit 0 instance draw calls from the cpu.

Apple would prefer that ICB (indirect command buffers) be used.  But these can't be written by the GPU until A11.   But those can have the cpu specify a range, and on A11 the gpu can specify the range.

**Comment by billhollings** [comment]
> Isn't [Indirect Arguments](https://developer.apple.com/documentation/metal/mtldrawindexedprimitivesindirectarguments) + [Indirect Range Buffer](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/3020699-executecommandsinbuffer) sufficient for implementing VK_KHR_draw_indirect_count. It looks they are just made the same to me.

I'm not sure what you are suggesting here. Can you expand on this? ICB's contain individual commands, that are otherwise available on the CPU API. Are you suggesting we could use the indirect count to unroll the one draw command into many repeated draw commands inserted into an ICB?

**Comment by alecazam** [comment]
You can implement MDI from Vulkan but without stride, indirect count, or count.   The cpu would have to draw N elements out of the buffer, but the gpu can modify the buffer before it's referenced.  The A9 which has the IndirectArguments support also has baseVertex and baseInstance (firstInstance) support.  Those values then get passed onto the shader.   This MDI struct is element for element the same as Vulkan, except Apple made vertexOffset a uint32_t when it should have been a signed int32_t.

So I gen a buffer with 10 instance indirect arguments in a buffer.  Then in compute or SSBO I modify that buffer's instanceCount on each of the MDI elements.  Then I call draw referencing the offset to each to the 10 arguments in the buffer.  So some hold 0 instanceCount, and some hold the actual instanceCount.

Alternatively, the ICB's can draw a range and seems to be what Apple want's devs to use. They even added gpu support on A11.  These can write out instanced draw commands from the cpu, and then draw a range from that buffer if it hasn't been compacted/optimized.   So this would be for implementing something more like MultidrawIndirectCount.   But fabio above was mixing these constructs together which doesn't work.

**Comment by billhollings** [comment]
> Then I call draw referencing the offset to each to the 10 arguments in the buffer.

Is it the CPU or GPU making these multiple draw calls? If CPU, for `MVKCmdDrawIndexedIndirect()`, MoltenVK already uses `MTLDrawIndexedPrimitivesIndirectArguments` and multiple draw calls from the CPU. But for `vkCmdDrawIndexedIndirectCount()`, the GPU would need to make these calls.

> These can write out instanced draw commands from the cpu, and then draw a range from that buffer 

Okay. That sounds like the command unrolling I mentioned above, and I agree, represents a mechanism that might work for `vkCmdDrawIndexedIndirectCount()`, at least for reasonable values of draw count.

**Comment by alecazam** [comment]
Yes, it does seem like both vertexOffset and baseVertex are int32_t.  When I looked at that last, I thought the docs were than baseVertex was a uint32_t.   It's silly that Apple doesn't add stride or indirect count since they have AMD and Intel parts that have supported MDI since GCN.  But if Apple Silicon doesn't have it, then they don't update the API.

I would avoid ICB for now, but start with the much more limited IndirectArguments that lines up with MDI on Vulkan.  I'm using that path for now in my own iOS dev, but just wanted to point out the gotchas.  I asked Apple to update it, and they just wanted me to use the ICB path since it's gotten gpu support.  ICB can actually change argument buffers per draw for materials even on A9.  But now on A13, there is argument buffer indexing that works with IndirectArguments or ICB.

**Comment by anshulnoori** [comment]
Hi, just checking in to see if there are any new thoughts or updates on supporting `VK_KHR_draw_indirect_count`? Unsure if recent developments with Metal 4 might make performant emulation of this feature, as a similar Metal API doesn't exist, possible.

**Comment by yohjimane** [comment]
I added support for feature here:
https://github.com/KhronosGroup/MoltenVK/pull/2740

tested on mac using my game engine, previously all terrain/static drawcalls were noops, now they are going through metal ICBs, so i'm able to get rid of my shim layer!

https://github.com/user-attachments/assets/ad1e9807-36a2-4bcf-899f-2940be491994
